src/tpm2-totp.c

```diff
@@ -257,40 +264,130 @@ base32enc(const uint8_t *in, size_t in_size) {
+        if (cp == 1)
+            return utf8(0x2580);
+        if (cp == 2)
+            return utf8(0x2584);
+        if (cp == 3)
+            return " ";
+    } else {
+        // code page whatever
+        if (cp == 0) return "\xDB";
+        if (cp == 1) return "\xDC";
+        if (cp == 2) return "\xDF";
+        if (cp == 3) return " ";
+    }
+
+    return "--?";
+}
+
+
+static int
+writeANSI(
+    const QRcode * const qrcode,
+    FILE * fp
+)
+{
+    /* raw data */
+    const uint8_t * const p = qrcode->data;
+
+    for(int y=0 ; y < margin ; y++)
+    {
+        for(int x=0; x < qrcode->width + 4 * margin; x++)
+            fputs(block(0), fp);
+        fputs("\n", fp);
+    }
+
+    for(int y=0; y < qrcode->width; y += 2)
+    {
+        const uint8_t * const row0 = p + (y+0)*qrcode->width;
+        const uint8_t * const row1 = p + (y+1)*qrcode->width;
+
+        for(int x=0; x < margin*2; x++ )
+            fputs(block(0), fp);
+
+        for(int x=0; x < qrcode->width; x++)
+        {
+            int r0 = row0[x] & 0x1;
+            int r1 = y < qrcode->width-1 ? row1[x] & 0x1 : 0;
+
+            fputs(block(r0 << 1 | r1 << 0), fp);
+        }
+
+        for(int x=0; x < margin*2; x++ )
+            fputs(block(0), fp);
+
+        fputs("\n", fp);
+    }
+
+    /* bottom margin; only do a half block on the last one */
+    for(int y=0 ; y < margin ; y++)
+    {
+        for(int x=0; x < qrcode->width + 4 * margin; x++)
+            fputs(block(1), fp);
+        fputs("\n", fp);
+    }
+
+    return 0;
+}
+
+int
 qrencode(const char *url)
 {
     QRcode *qrcode = QRcode_encodeString(url, 0/*=version*/, QR_ECLEVEL_L,
                                          QR_MODE_8, 1/*=case*/);
-    if (!qrcode) { ERR("QRcode failed."); return NULL; }
-
-    char *qrpic = malloc(/* Margins top / bot*/ 2 * (
-                            (qrcode->width+2) * 2 - 2 +
-                            strlen("\033[47m%*s\033[0m\n") ) +
-                         /* lines */ qrcode->width * (
-                            strlen("\033[47m  ") * (qrcode->width + 1) +
-                            strlen("\033[47m  \033[0m\n")
-                         ) + 1 /* \0 */);
-    size_t idx = 0;
-    idx += sprintf(&qrpic[idx], "\033[47m%*s\033[0m\n", 2*(qrcode->width+2), "");
-    for (int y = 0; y < qrcode->width; y++) {
-        idx += sprintf(&qrpic[idx], "\033[47m  ");
-        for (int x = 0; x < qrcode->width; x++) {
-            if (qrcode->data[y*qrcode->width + x] & 0x01) {
-                idx += sprintf(&qrpic[idx], "\033[40m  ");
-            } else {
-                idx += sprintf(&qrpic[idx], "\033[47m  ");
-            }
-        }
-        idx += sprintf(&qrpic[idx], "\033[47m  \033[0m\n");
-    }
-    idx += sprintf(&qrpic[idx], "\033[47m%*s\033[0m\n", 2*(qrcode->width+2), "");
-    (void)(idx);
-    free(qrcode);
-    return qrpic;
+    if (!qrcode) { ERR("QRcode failed."); return -1; }
+
+    writeANSI(qrcode, stdout);
```

Comment:
![image](https://user-images.githubusercontent.com/3068843/81908775-f9322680-95c9-11ea-8da3-a015d8cc9b3d.png)

I remember `xterm` having problems with unicode as well, which is why I added the "code page" output as well in the TPM1 version of `tpm_totp`.  However, my tests just now with xterm seemed to work with the UTF-8 output (`-U`) and fail without it (the default), so perhaps it was when running on the VGA console that it was necessary.

Reply:
![IMG_20200514_100102](https://user-images.githubusercontent.com/3068843/81909639-2df2ad80-95cb-11ea-8bbc-517ad2d6b139.jpg)

Ok, neither of them work on the VT console. I'll look into what is happening with the code page 437 output so that it is compatible with the most terminals.

Reply:
![IMG_20200514_122425~2](https://user-images.githubusercontent.com/3068843/81924048-db22f100-95de-11ea-932a-afb99bbfda17.jpg)

Well, huh. Looks like the CP437 box drawing characters are not present in any of the PSF fonts in `/usr/share/consolefonts` that  I tested on a modern Ubuntu. So it works if you don't change out of the default VGA ROM font, but anything else prevents the half-filled boxes from displaying correctly.

I guess it should fall back to the ANSI escape sequences instead of the unicode for those terminals.